src/kernel/mem/kheap.c

```diff
@@ -6,86 +6,194 @@
-    return NULL;
+uint64_t compute_kheap_end() {
+    return (uint64_t)kernel_heap_end + kernel_heap_end->size + sizeof(KHeapMemoryNode);
+}
+
+
+void kfree(void *ptr) {
+    // Before doing anything let's check that the address provided is valid: not null, and within the heap space
+    if(ptr == NULL) {
+        return;
+    }
+
+    if ( (uint64_t) ptr < (uint64_t) kernel_heap_start && (uint64_t) ptr > (uint64_t) kernel_heap_end) {
+        return;
+    }
+
+    // Now we can search for the node containing our address
+    KHeapMemoryNode *current_node = kernel_heap_start;
+    while( current_node != NULL ) {
+        if( ((uint64_t) current_node + sizeof(KHeapMemoryNode)) == (uint64_t) ptr) {
+            current_node->is_free = true;
+            uint8_t available_merges = can_merge(current_node);
+ 
+            if( available_merges & MERGE_RIGHT ) {
+                merge_memory_nodes(current_node, current_node->next);
+            }
+           
+            if( available_merges & MERGE_LEFT ) {
+                merge_memory_nodes(current_node->prev, current_node);
+            }
+            return;
+            
+        }
+        current_node = current_node->next;
+    }
+}
 
+#ifdef _TEST_
+uint8_t get_kheap_size(KHeapMemoryNode *heap_start) {
+    KHeapMemoryNode *cur_node = heap_start;
+    uint8_t size = 0;
+    while( cur_node != NULL ) {
+        size++;
+        cur_node = cur_node->next;        
+    }
+    return size;
 }
+#endif
 
-void kfree(void *ptr){
+uint8_t can_merge(KHeapMemoryNode *cur_node) {
+    // This function checks if the current node can be merged to both left and right
+    // There return value is a 2 bits field: bit #0 is set if the node can be merged right
+    // bit #1 is set if the node can be merged left. Bot bits set means it can merge in both diections
+    KHeapMemoryNode *prev_node = cur_node->prev;
+    KHeapMemoryNode *next_node = cur_node->next;
+    uint8_t available_merges = 0;
+    if( prev_node != NULL && prev_node->is_free ) {
+        uint64_t prev_address = (uint64_t) prev_node + sizeof(KHeapMemoryNode) + prev_node->size;
+        if ( prev_address == (uint64_t) cur_node ) {
+            available_merges = available_merges | MERGE_LEFT;
+        }
+    }
+    if( next_node != NULL && next_node->is_free ) {
+        uint64_t next_address = (uint64_t) cur_node + sizeof(KHeapMemoryNode) + cur_node->size;
+        if ( next_address == (uint64_t) cur_node->next ) {
+            available_merges = available_merges | MERGE_RIGHT;
+        }
+
+    }
+
+    return available_merges;
 }
+void merge_memory_nodes(KHeapMemoryNode *left_node, KHeapMemoryNode *right_node) {
+    if(left_node == NULL || right_node == NULL) {
+        return;
+    }
+    if(((uint64_t) left_node +  left_node->size + sizeof(KHeapMemoryNode)) == (uint64_t) right_node) {
+        //We can combine the two nodes:
+        //1. Sum the sizes
+        left_node->size = left_node->size + right_node->size;
```

Comment:
You'll also want to add the bytes used by the heap node here too, since that node no longer exists (and is no longer using that memory), it's now part of the memory owned by the left node.